Visualizing chunks tutorial (#144)
timescale/docs: timescaledb/tutorials/building-a-custom-timescaledb-dashboard.md

> # Building a custom TimescaleDB dashboard

Author: should I rename all the files to make the new title the standard?

> Databases don’t have to be black boxes. With a few simple steps, you can create your own custom visualizations and dashboards to monitor your TimescaleDB instance. TimescaleDB is PostgreSQL with superpowers and makes creating custom dashboards for your data easy. With TimescaleDB you have the full canvas of PostgreSQL monitoring capabilities to hook into. Of course, you can always use commercial tools to monitor TimescaleDB, just as you can with PostgreSQL. But custom dashboards give you the utmost flexibility. Fortunately, it’s not difficult to build these kinds of dashboards.

Author: Thank you @Loquacity!

> using the full functionality of PostgreSQL monitoring

I'm a bit confused with the `monitoring` part.

We will also make the visualization interactive by enabling you to click on a chunk (one of the dots on the screen) and compress or uncompress it manually.

## Create internal views in TimescaleDB to obtain metrics

To make it easier to build our application, we had to create some new functions and views to do the following:

* Extract information from chunks like name and time range.
* Get extra details about what chunks are compressed or not.
* Get compression stats and fetch what is the size after compression.

### Extract information from chunks

To address the first item, we can use the timescaledb_information.chunks view that the TimescaleDB extension already provides.

Here is a simple query to get the time-series range of each chunk:

```sql
 SELECT hypertable_schema,
    hypertable_name,
    chunk_name,
    range_start,
    range_end
 FROM timescaledb_information.chunks LIMIT 1;
```

> Sample row vertically output to explore:

Author: Not sure what is your question here. Basically, I split the query from the output. And the results are aligned vertically not horizontally as the regular queries output.

after_compression_index_bytes  | 16384
after_compression_toast_bytes  | 98304
after_compression_total_bytes  | 122880
node_name                      |
-[ RECORD 2 ]------------------+----------------------
chunk_schema                   | _timescaledb_internal
chunk_name                     | _hyper_6_880_chunk
compression_status             | Uncompressed
before_compression_table_bytes |
before_compression_toast_bytes |
before_compression_total_bytes |
after_compression_table_bytes  |
after_compression_index_bytes  |
after_compression_toast_bytes  |
after_compression_total_bytes  |
node_name                      |
```

### Get compression stats and size

One detail you might notice is that when the chunk is uncompressed, it does not bring the size of the chunk. For the uncompressed data, we need to fall back to the `chunks_detailed_size` function, passing a specific hypertable name as a parameter:

```sql
tsdb=> SELECT * FROM chunks_detailed_size('conditions');
-[ RECORD 1 ]+----------------------
chunk_schema | _timescaledb_internal
chunk_name   | _hyper_6_853_chunk
table_bytes  | 8192
index_bytes  | 40960
toast_bytes  | 98304
total_bytes  | 147456
node_name    |
```

> From this function we can use the `total_bytes` when we know that the chunk is not compressed.

Reviewer: ```suggestion
You can use the `total_bytes` information in this function to see that the chunk is uncompressed.
```

### Building views for our TimescaleDB metrics

> Now that we know where to get all of the data we need to drive the visualization, it’s time to join it together in a view that can be queried using SQL (and eventually, our application).

Reviewer: ```suggestion
Now that you know how to gather all of the data you need to drive the visualization, it’s time to join it together in a view that can be queried using SQL (and eventually, our application).
```

Author: thanks a lot @Loquacity for all the suggestions from `we -> you` I'm still getting used to who is who here 😄

```sql
CREATE OR REPLACE VIEW chunks_with_compression AS
SELECT DISTINCT ch.chunk_name,
                ccs.chunk_schema,
                ch.hypertable_schema,
                ch.hypertable_name,
                ch.range_start,
                ch.range_end,
                COALESCE(ccs.before_compression_total_bytes, NULL, cds.total_bytes) AS before_compression_total_bytes,
                ccs.after_compression_total_bytes
FROM (
 SELECT hypertable_schema,
    hypertable_name,
    chunk_name,
    range_start,
    range_end
 FROM  timescaledb_information.chunks) AS ch
  LEFT OUTER JOIN LATERAL chunk_compression_stats(ch.hypertable_name::regclass) ccs
    ON ch.chunk_name = ccs.chunk_name
  LEFT OUTER JOIN LATERAL chunks_detailed_size(ch.hypertable_name::regclass) cds
    ON ccs.chunk_schema = cds.chunk_schema
    AND ch.chunk_name = cds.chunk_name;
```

> Using the name of a random chunk from the hypertable, we can query this view and see that we now get all of the information we need, including the time range of the chunk, it's size before compression, and the size after compression if applicable.

Reviewer: ```suggestion
To test, use the name of a random chunk from the hypertable to query this view and check that you get all of the information you need. You should see the time range of the chunk, the hypertable information, and its size before and after compression.
```

> Notice in this sample chunk from our test dataset that `before_compression_total_bytes` is 10 times bigger than `after_compression_total_bytes`. We save more than 90% of disk space with compression!

Reviewer: ```suggestion
In this example chunk, the `before_compression_total_bytes` is ten times bigger than `after_compression_total_bytes`. Compression saved more than 90% of disk space!
```

```sql
SELECT * FROM  chunks_with_compression;
...
-[ RECORD 96 ]-----------------+-----------------------
chunk_name                     | _hyper_2_37_chunk
chunk_schema                   | _timescaledb_internal
hypertable_schema              | public
hypertable_name                | conditions
range_start                    | 2021-05-27 00:00:00+00
range_end                      | 2021-06-03 00:00:00+00
before_compression_total_bytes | 90112
after_compression_total_bytes  | 8192
```

## Setting up your database

> In this project, we use data generated by our tutorial on [Simulating IoT sensor data][iot-tutorial]. This data will result in a simple schema and data that mimics a number of IoT sensors with information on time, device, and temperature.

Reviewer: ```suggestion
In this example, we are using data generated by our [Simulating IoT sensor data][iot-tutorial] tutorial. This data results in a simple schema and data that mimics a number of IoT sensors with information on time, device, and temperature.
```

> If you follow the tutorial, you will get a table named conditions, which stores the temperature of fake devices over time.

Reviewer: ```suggestion
By following the tutorial, you have a table named `conditions`, which stores the temperature of example devices over time.
```

> Here’s a set of commands to create the table and generate some sample data:

Reviewer: ```suggestion
Use these commands to create the table and generate some sample data:
```

```sql
CREATE TABLE conditions (
      time TIMESTAMPTZ NOT NULL,
      device INTEGER NOT NULL,
      temperature FLOAT NOT NULL,
      PRIMARY KEY(time, device)
);

SELECT * FROM create_hypertable('conditions', 'time', 'device', 3);

INSERT INTO conditions
  SELECT time, (random()*30)::int, random()*80 - 40
  FROM generate_series(TIMESTAMP '2020-01-01 00:00:00',
                       TIMESTAMP '2020-01-01 00:00:00' + INTERVAL '1 month',
             INTERVAL '1 min') AS time;

```

> ## Connecting to our database and retrieving metrics

Reviewer: ```suggestion
## Connecting to the database and retrieving metrics
```

> One of the most trivial aspects of a backend application is to protect the database and expose only the information that is accessible to the authorized user. The Hasura GraphQL Engine solves this issue providing GraphQL APIs over new or existing PostgreSQL databases. Allowing you to create permission rules and dynamically expand your database resources.

Reviewer: ```suggestion
When you write a backend application, you need to protect the database and expose only the required information to an authorized user. The Hasura GraphQL Engine does this by providing GraphQL APIs over new or existing PostgreSQL databases. This allows you to create permission rules and dynamically expand your database resources.
```

> Now that we have a database example, we can jump into the [Hasura cloud][hasura-cloud] to connect the resources that we want to expose through the GraphQL.
> 
> We chose Hasura because of the flexibility to simply connect our TimescaleDB database and quickly expose the tables, views, and functions we needed in a few clicks. We’re not going to dive into all the steps to set up a new data source on Hasura as the wizard will guide you there.

Reviewer: ```suggestion
When you have your sample database set up, you can use the [Hasura cloud][hasura-cloud] to connect the resources that we want to expose through GraphQL. Hasura is a good option because it connects to our TimescaleDB database and quickly exposes the tables, views, and functions you need. For more information about setting up a new data source on Hasura, check out their wizard.
```

> We’re going to use two types of operations:
> 
> * Queries and subscriptions: watch a specific query and keep pulling data updates to the client. We’re going to "subscribe" to the chunks metadata’.
> * Mutation: convention for operations that are writing data. We’ll map the compression and decompression actions as mutations.

Reviewer: ```suggestion
We’re going to use two types of operations:

* Queries and subscriptions: watch a specific query and keep pulling data updates to the client. In this example, you subscribe to the chunks' metadata.
* Mutation: convention for operations that write data. In this example, you map the compression and decompression actions as mutations.
```

"convention"?

Author: You rock @Loquacity! I'm learning a lot from your reviews and suggestions ❤️

### Queries and subscriptions

> Hasura has great flexibility to plug any resource and offer it as a query or a subscription. For this example, we’ll be mapping the *chunks_with_compression* view we created earlier as a GraphQL resource. It can be consumed as a query or subscription.
> 
> ![Tracking a SQL view on Hasura cloud](https://assets.timescale.com/docs/images/tutorials/visualizing-compression/hasura-cloud-track-view.png)
> 
> We need to map those mutations to compress and decompress a chunk. To accomplish that, we need to take a look at how Hasura types work.

Reviewer: ```suggestion
Hasura allows you to attach any resource and offer it as a query or a subscription. In this example, you map the `chunks_with_compression` view you created earlier as a GraphQL resource, so it can be consumed as a query or subscription. You can then map the changes, or mutations, as you compress and decompress a chunk. This image describes a SQL view is tracked on Hasura:

![Tracking a SQL view on Hasura cloud](https://assets.timescale.com/docs/images/tutorials/visualizing-compression/hasura-cloud-track-view.png)
```

### Mutations

> Hasura can easily map custom types that come from table structures. To create the necessary mutations, our functions should also return types that inherit from table structures.
> 
> To easily create a new structure of the table from a query, we can call the query with limit 0:

Reviewer: ```suggestion
Hasura can map custom types that come from table structures. To create the necessary mutations, functions need to return types that inherit from table structures. To create a new structure of the table from a query, call the query with limit 0:
```

#### Compress chunk mutation

```sql
CREATE TABLE compressed_chunk AS
SELECT compress_chunk((c.chunk_schema ||'.' ||c.chunk_name)::regclass)
FROM   timescaledb_information.chunks c
WHERE  NOT c.is_compressed limit 0;
```

> Hasura needs some function to be tracked as a mutation. In this case, let's create the function to just rewrap the default compress_chunk from the TimescaleDB extension.
> 
> Now, we can return the "compressed_chunk" in our function that will compress the chunk:

Reviewer: ```suggestion
Hasura needs a function to be tracked as a mutation. Create a function to rewrap the default `compress_chunk` from the TimescaleDB extension, and return the "compressed_chunk" in a function that compresses the chunk:
```

I'm not sure this makes sense.

```sql
CREATE OR REPLACE FUNCTION compress_chunk_named(varchar) returns setof compressed_chunk AS $$
  SELECT compress_chunk((c.chunk_schema ||'.' ||$1)::regclass)
  FROM   timescaledb_information.chunks c
  WHERE  NOT c.is_compressed
  AND    c.chunk_name = $1 limit 1
$$ LANGUAGE SQL VOLATILE;
```

> Note that the function adds an extra where clause to not compress what is already compressed.

Reviewer: ```suggestion
Note that the function adds an extra `where` clause so that it does not compress a chunk that is already compressed.
```

![Tracking compress chunk mutation on Hasura cloud](https://assets.timescale.com/docs/images/tutorials/visualizing-compression/hasura-cloud-compress-chunk-mutation.png)

#### Decompress chunk mutation

> We'll need a similar function for the decompression:

Reviewer: ```suggestion
You also need a similar function for decompression:
```

```sql
CREATE OR REPLACE FUNCTION decompress_chunk_named(varchar) returns setof compressed_chunk AS $$
  SELECT decompress_chunk((c.chunk_schema ||'.' ||$1)::regclass)
  FROM   timescaledb_information.chunks c
  WHERE  c.is_compressed
  AND    c.chunk_name = $1 limit 1
$$ LANGUAGE SQL VOLATILE;
```

> Now, the next step is to jump into the Hasura cloud and connect the database as a new data source.
> 
> In the data panel, after setting up the PostgreSQL URI of your database, you can track each function as a query or mutation in a single click. Here is an example of compress_chunk_named function:
> 
> In our case, the subscription goes to the chunks_with_compression, and here is what it looks like:
> 
> You should also track decompress_chunk_named and compress_chunk_named as GQL mutations with a single argument.

Reviewer: ```suggestion
The next step is to go to the Hasura cloud and connect the database as a new data source. In the data panel, set up the PostgreSQL URI of your database, and then you can track each function as a query or mutation. This is an example of the `compress_chunk_named` function. In our case, the subscription goes to the  `chunks_with_compression` function. You can also track `decompress_chunk_named` and `compress_chunk_named` as GQL mutations with a single argument.
```
Missing images?

Author: oh yes, the images were a bit repetitive, then we have one less image here. Let me double-check in the final reading and I'll re-add if needed. As you rewrote these pieces now it looks awesome 🙇

## Build the front-end visualization

> You can consult [the GitHub repo][repo-example] for the full code of our front-end application. The front-end application will connect to the Hasura GraphQL layer we created earlier, which, in turn, connects to our TimescaleDB database to retrieve information about chunks and compression status. The front-end application will then render the "dots" for our visualization.
> 
> As a summary, in the front-end, we will do the following:
> 
> 1. Subscribe to the API via GraphQL
> 2. Create an SVG component
> 3. Iterate over all the chunks and add circles in the previous component
> 4. Style the circle and add events to interact with it

Reviewer: ```suggestion
For the full code of our front-end aplpication, see our [GitHub repo][repo-example]. The front-end application connects to the Hasura GraphQL layer you created, then connects to the TimescaleDB database to retrieve information about chunks and compression status. The front-end application then renders the image for the visualization.

As a summary, the front-end:

1. Subscribes to the API with GraphQL
2. Creates an SVG component
3. Iterates over all the chunks, and adds circles in the previous component
4. Styles the circle and adds events to interact with the image
```

## Summary

> TimescaleDB is a powerful relational database for time-series data. But TimescaleDB is also built on top of PostgreSQL, the fastest growing database in the world. As a result, you can not only leverage the full spectrum of tools and dashboards available for PostgreSQL, but you can also build your own custom tools and dashboards to connect to PostgreSQL.
> 
> The easiest way to get a TimescaleDB instance is to [try for free][timescale-signup] using our hosted service. You can also [download TimescaleDB for free][timescale-install] and run locally or in your own cloud infrastructure.
> 
> You can get the full code for this project from [this GitHub repo][repo-example].
> 
> We hope you find new ways to explore your data and make your decisions smarter and data-driven. If you get any interesting results or have any questions about this tutorial, drop us a line on our [community Slack channel][timescale-slack].

Reviewer: ```suggestion
You can use TimescaleDB to build your own custom tools and dashboards to connect to PostgreSQL. Get a TimescaleDB instance to [try for free][timescale-signup] using our hosted service. You can also [download TimescaleDB for free][timescale-install] and run it locally or in your own cloud infrastructure.

You can get the full code for this project from [this GitHub repo][repo-example].

We hope you find new ways to explore your data and make your decisions smarter and data-driven. If you get any interesting results or have any questions about this tutorial, drop us a line on our [community Slack channel][timescale-slack].
```

This doesn't read like a summary, more of a conclusion?

Author: Yes, you're right. When we initially build it, it was a blog post, but now we need to focus more on the Summary perspective. I'll be reworking on this part.

Author: I reworked it here, please review it again 🙇
